src/hotspot/share/opto/parseHelper.cpp

```diff
@@ -650,12 +635,44 @@ Node* PEAState::materialize(GraphKit* kit, Node* var) {
 
   const TypeOopPtr* oop_type = var->as_Type()->type()->is_oopptr();
   Node* objx = kit->materialize_object(alloc, oop_type);
-  VirtualState* virt = static_cast<VirtualState*>(get_object_state(alloc));
 
   // we save VirtualState beforehand.
   escape(alloc, objx, true);
-  replace_in_map(kit, var, objx);
   pea->add_alias(alloc, objx);
+
+  if (virt->lockcnt() > 0 && GenerateSynchronizationCode) {
+    if (PEAVerbose) {
+      tty->print_cr("materializing an object with unbalanced monitor");
+    }
+
+    int mon_id = 0;
+    JVMState* jvms = kit->jvms();
+    int cnt = 0;
+    // It's possible that the locked moninitor is not in the youngest JVMState,
+    // so we have to follow the stacktrace to discover them all.
+    //
+    // PEA Materialization steals those monitors from the original object. Here is the scheme:
+    // 1. unlock the original object.
+    // 2. lock the materialized object.
+    // 3. backfill the obj for Monitor 'obj|box' pair when Parse translates 'monitor-exit'.
+    // 4. split Phi-Unlock, so EA and ME track them individually.
```

Comment:
Are we splitting in this function? I don't see the code for this. I think we only need to split when we see unlock node

Reply:
we don't split Phi-Unlock in materialization. I will update the comment.